### DISCUSSION TITLE
Consequeneces of upgrading Strimzi on existing Kafka Clusters

### DISCUSSION THREAD

**Discussion by 282ori** [strimzi/strimzi-kafka-operator#7104] [Q&A]
Hi @scholzj,
I have a general question:
Lets assume I have some Kafka Clusters and I'm upgrading Strimzi to higher version.
If I'm not upgrading the Kafka Clusters (not changing the version in the Kafka CRs),
could be any effects on these clusters?
I mean if it's possible that these clusters won't work properly until I change their version.
I know that the Strimzi won't watch the CRs but I want to know if the clusters will be fine for sure.
Thanks

**Comment by scholzj** [comment]
> I know that the Strimzi won't watch the CRs but I want to know if the clusters will be fine for sure.

In general, Strimzi will watch the same CRs before and after the upgrade (assuming you did not changed the confguration).

> Lets assume I have some Kafka Clusters and I'm upgrading Strimzi to higher version.
If I'm not upgrading the Kafka Clusters (not changing the version in the Kafka CRs),
could be any effects on these clusters?

Each Strimzi version supports selected set of Kafka versions. So you need to distinguish two situations:
1) The Kafka cluster uses a Kafka version supported by both old and new Strimzi versions
2) The Kafka cluster uses a Kafka version not supported by the new Strimzi version

In case of 1, the operator will keep using the Kafka version, but it will always roll the operands to the new container images (which even with the same Kafka version but various dependency updates, CVE fixes, scripts improvements etc.)

In case of 2, the operator will see the custom resource, but just throw an error about unsupported version instead of reconciling the cluster. In this case, it will basically not roll it and let it be alone.

> I mean if it's possible that these clusters won't work properly until I change their version.

Assuming this is the case of 2 from above, it will leave the cluster be as it was. But it will not operate it. So if the cluster break, it will be broken, it certs expire, it will be broken, etc. So it basically depends. But the operator it self will not break it, it will just leave it to its own and not operate it.

**Reply by 282ori** [reply]
Thanks, I have a few more questions:

1. If I apply new CRDs, In which cases do I have to change the current Kafka CR's to the new CRD scheme in order to ensure that the cluster operator keeps reconciling the Kafka CR's?

2. my current Strimzi supports for example kafka versions: 2.5.0, 2.6.0 and 2.7.0 and I have an existing Kafka with Kafka.spec.kafka.version: 2.7.0 
If I upgrade Strimzi to next version and it supports kafka versions: 2.7.0, 2.8.0, 2.9.0,
would the Cluster operator replace the current image of the cluster from the old image (of the old Strimzi version) to the new image of the new operator? (and perform rolling update on the cluster's brokers)

Thanks

**Reply by scholzj** [reply]
Re 1) I don't think so. The CRD schema is not really related to the Kafka version. If you would have some invalid CR, the operator should normally just throw some errors, but not break your cluster. It will again simply stop reconciling it. But the Kafka pods etc. should be undisrupted at that point (once again, since the operator will not operate them, certificates might expire, the cluser might break for another reasons etc., but the fact that the operator does not operate them is not something what breaks them immediately). Also, keep in mind that within the same API version, the CRDs are backwards compatible. The only recent (almost year and half ago IIRC) breaking change as between 0.22 and 0.23 releases which required a conversion. But in most upgrades, no CRD / CR conversion is needed since your CR is still valid.

Re 2) Yes, as I said -> if the new version of the operator supports the Kafka version you use, it will still roll the Kafka cluster to use the new images. This is required because the images (apart from CVE fixes etc.) might also contain some new functionality, enhancements etc. But your Kafka version will be still the same.
